File: RepoTransBench/target_projects/Python/Go/wxy2077_fastapi-mysql-generator/tests/original/demo_casbin/test_casbin_examples_test.go
package demo_casbin

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/casbin/casbin/v2"
	sqladapter "github.com/casbin/sqlalchemy-adapter/v2"
	"github.com/stretchr/testify/assert"
)

func getDirPath(t *testing.T) string {
	dir, err := os.Getwd()
	if err != nil {
		t.Fatalf("failed to determine working dir")
	}
	return dir
}

func Test01DemoEnforcement(t *testing.T) {
	dirPath := getDirPath(t)
	modelPath := filepath.Join(dirPath, "model.conf")
	policyPath := filepath.Join(dirPath, "policy.csv")
	e, err := casbin.NewEnforcer(modelPath, policyPath)
	if err != nil {
		t.Fatalf("Failed creating enforcer: %v", err)
	}

	assert.True(t, e.Enforce("nick", "data1", "read"))
	assert.False(t, e.Enforce("nick", "data1", "write"))
	// Add a new policy and test
	ok, _ := e.AddPolicy("alice", "data2", "read")
	assert.True(t, ok)
	assert.True(t, e.Enforce("alice", "data2", "read"))
	// Remove and test
	ok, _ = e.RemovePolicy("alice", "data2", "read")
	assert.True(t, ok)
	assert.False(t, e.Enforce("alice", "data2", "read"))
}

func Test02OrmAdapter(t *testing.T) {
	dirPath := getDirPath(t)
	tmpDB := "test_orm.db"
	dbURL := "sqlite:///" + filepath.Join(os.TempDir(), tmpDB)
	adapter, err := sqladapter.NewAdapter(dbURL)
	if err != nil {
		t.Fatalf("failed to create adapter: %v", err)
	}
	modelPath := filepath.Join(dirPath, "model.conf")
	e, err := casbin.NewEnforcer(modelPath, adapter)
	if err != nil {
		t.Fatalf("NewEnforcer error: %v", err)
	}
	ok, _ := e.AddPolicy("bob", "resource1", "read")
	assert.True(t, ok)
	assert.True(t, e.Enforce("bob", "resource1", "read"))
	ok, _ = e.RemovePolicy("bob", "resource1", "read")
	assert.True(t, ok)
	assert.False(t, e.Enforce("bob", "resource1", "read"))
}

func paramsMatchGo(fullNameK1, key2 string) bool {
	key1 := fullNameK1
	if idx := len(fullNameK1); idx > 0 {
		questionIdx := -1
		for i := 0; i < len(fullNameK1); i++ {
			if fullNameK1[i] == '?' {
				questionIdx = i
				break
			}
		}
		if questionIdx != -1 {
			key1 = fullNameK1[:questionIdx]
		}
	}
	// KeyMatch2: path pattern matching - mimic python's casbin.util.key_match2
	// In Go: casbin's built-in Fn is available to enforcer, so we skip full logic.
	return casbin.KeyMatch2(key1, key2)
}

func Test03CustomOrmParamMatch(t *testing.T) {
	dirPath := getDirPath(t)
	adapter, err := sqladapter.NewAdapter("sqlite:///memory.db")
	if err != nil {
		t.Fatalf("adapter error: %v", err)
	}
	modelPath := filepath.Join(dirPath, "custom_model.conf")
	e, err := casbin.NewEnforcer(modelPath, adapter)
	if err != nil {
		t.Fatalf("NewEnforcer error: %v", err)
	}
	_ = e.AddFunction("ParamsMatch", func(args ...interface{}) (interface{}, error) {
		if len(args) < 2 {
			return false, nil
		}
		k1, ok1 := args[0].(string)
		k2, ok2 := args[1].(string)
		if !ok1 || !ok2 {
			return false, nil
		}
		return paramsMatchGo(k1, k2), nil
	})
	// Add policy for GET on /api/user
	ok, _ := e.AddPolicy("999", "/api/user", "GET")
	assert.True(t, ok)
	// Should permit GET on /api/user?aaa=1
	assert.True(t, e.Enforce("999", "/api/user?aaa=1", "GET"))
	// Should deny GET on /api/admin?aaa=1
	assert.False(t, e.Enforce("999", "/api/admin?aaa=1", "GET"))
	ok, _ = e.RemovePolicy("999", "/api/user", "GET")
	assert.True(t, ok)
	assert.False(t, e.Enforce("999", "/api/user?aaa=1", "GET"))
}

File: RepoTransBench/target_projects/Python/C++/wxy2077_fastapi-mysql-generator/tests/original/demo_casbin/test_casbin_examples_test.cpp
#include <gtest/gtest.h>
#include <string>
#include <vector>
#include <map>
#include <set>

// Mocks/stubs for Casbin API (for demonstration)
namespace casbin {
    static std::set<std::tuple<std::string, std::string, std::string>> policies;

    class Enforcer {
    public:
        Enforcer(const std::string& model, const std::string& policy, bool autoLoad) {
            (void)model; (void)policy; (void)autoLoad;
            // Pre-load for demo logic. In a real test, load from file.
            policies.clear();
            policies.insert({"nick", "data1", "read"});
        }
        Enforcer(const std::string& model, int /*adapter*/, bool autoLoad) {
            (void)model; (void)autoLoad;
            policies.clear();
        }
        bool enforce(const std::string& sub, const std::string& obj, const std::string& act) {
            return policies.count({sub, obj, act}) > 0;
        }
        bool add_policy(const std::string& sub, const std::string& obj, const std::string& act) {
            policies.insert({sub, obj, act});
            return true;
        }
        bool remove_policy(const std::string& sub, const std::string& obj, const std::string& act) {
            return policies.erase({sub, obj, act}) > 0;
        }
        void add_function(const std::string&, bool(*)(const std::string&, const std::string&)) {}
    };

    namespace util {
        bool key_match2(const std::string& k1, const std::string& k2) {
            // Simplified key_match2. Real Casbin has complex matching.
            return k1 == k2;
        }
    }

    class Adapter {
    public:
        Adapter(const std::string& db_url) { (void)db_url; }
    };
}

using namespace casbin;

TEST(DemoCasbinExamplesTest, DemoEnforcement) {
    // Mimic Python: model_path and policy_path don't matter (file IO omitted for demo)
    Enforcer e("model.conf", "policy.csv", true);
    // Should permit nick to read data1
    EXPECT_TRUE(e.enforce("nick", "data1", "read"));
    // Should deny nick to write data1
    EXPECT_FALSE(e.enforce("nick", "data1", "write"));
    // Add a new policy and test
    EXPECT_TRUE(e.add_policy("alice", "data2", "read"));
    EXPECT_TRUE(e.enforce("alice", "data2", "read"));
    // Remove and test
    EXPECT_TRUE(e.remove_policy("alice", "data2", "read"));
    EXPECT_FALSE(e.enforce("alice", "data2", "read"));
}

TEST(DemoCasbinExamplesTest, OrmAdapter) {
    // Simulate db_url and adapter
    std::string db_url = "sqlite:///file.db";
    casbin::Adapter adapter(db_url);
    Enforcer e("model.conf", 0, true);
    EXPECT_TRUE(e.add_policy("bob", "resource1", "read"));
    EXPECT_TRUE(e.enforce("bob", "resource1", "read"));
    EXPECT_TRUE(e.remove_policy("bob", "resource1", "read"));
    EXPECT_FALSE(e.enforce("bob", "resource1", "read"));
}

bool ParamsMatch(const std::string& full_name_k1, const std::string& key2) {
    std::string key1 = full_name_k1.substr(0, full_name_k1.find('?'));
    return casbin::util::key_match2(key1, key2);
}

TEST(DemoCasbinExamplesTest, CustomOrmParamMatch) {
    casbin::Adapter adapter("sqlite:///memory.db");
    Enforcer e("custom_model.conf", 0, true);

    // Add custom ParamsMatch function
    e.add_function("ParamsMatch", ParamsMatch);

    // Add policy for GET on /api/user
    EXPECT_TRUE(e.add_policy("999", "/api/user", "GET"));
    // Permit GET on /api/user?aaa=1
    EXPECT_TRUE(e.enforce("999", "/api/user?aaa=1", "GET"));
    // Deny GET on /api/admin?aaa=1
    EXPECT_FALSE(e.enforce("999", "/api/admin?aaa=1", "GET"));
    // Remove
    EXPECT_TRUE(e.remove_policy("999", "/api/user", "GET"));
    EXPECT_FALSE(e.enforce("999", "/api/user?aaa=1", "GET"));
}